[JIT] preserve ignored function return value type

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25262 [JIT] preserve ignored function return value type**

Preserve the type of @ignore'd functions on serialization. Currently we first compile an @ignore'd function with it's annotated type when first compiling, but do not preserve it. This is important for being able to compile models with not-yet-supported features in JIT.

```
@torch.jit.ignore
def unsupported(x):
    return x

def foo():
   if not torch.jit._is_scripting():
      return torch.linear(...)
   else:
      return unsupported(...)
```

Differential Revision: [D17272443](https://our.internmc.facebook.com/intern/diff/D17272443)

Commit:
preserve ignored function return value type

eellison commented:
@zou3519 @driazati any thoughts on the name of the api ? or are we good with this

Commit:
Update on "[JIT] preserve ignored function return value type"


Preserve the type of @ignore'd functions on serialization. Currently we first compile an @ignore'd function with it's annotated type when first compiling, but do not preserve it. This is important for being able to compile models with not-yet-supported features in JIT.

```
@torch.jit.ignore
def unsupported(x):
    return x

def foo():
   if not torch.jit._is_scripting():
      return torch.linear(...)
   else:
      return unsupported(...)
```

Commit:
Update on "[JIT] preserve ignored function return value type"


Preserve the type of @ignore'd functions on serialization. Currently we first compile an @ignore'd function with it's annotated type when first compiling, but do not preserve it. This is important for being able to compile models with not-yet-supported features in JIT.

```
@torch.jit.ignore
def unsupported(x):
    return x

def foo():
   if not torch.jit._is_scripting():
      return torch.linear(...)
   else:
      return unsupported(...)
```

eellison commented:
@driazati Michael and I both think it would make more sense to insert an exception on the first compilation instead of waiting for save to occur. This also simplifies the code. 

I'm not sure if we want to deprecate the old `DROP_ON_EXPORT` api or if not drop it make an option that is clearer what we're doing, `DROP_ON_COMPILE` or something else

eellison commented:
@suo @driazati take a look? This is for use in torchvision, which doesn't build off of master it builds off of the nightly, so there's a0 bit of a lag in incorporating new changes...

driazati commented:
> insert an exception on the first compilation

The main use case is to support training and debugging. With `drop_on_compile`, training is still fine since with the new API all the training can happen in Python, then people `torch.jit.script()` when they are ready to export. The debugging use case would regress with `drop_on_compile` since right now you can have a function that is used to debug/log/visualize what the jit runtime is doing but gets removed on export but if it's dropped on compilation that can't be done.

eellison commented:
> > insert an exception on the first compilation
> 
> The main use case is to support training and debugging. With `drop_on_compile`, training is still fine since with the new API all the training can happen in Python, then people `torch.jit.script()` when they are ready to export. The debugging use case would regress with `drop_on_compile` since right now you can have a function that is used to debug/log/visualize what the jit runtime is doing but gets removed on export but if it's dropped on compilation that can't be done.

Okay that makes sense as a use case. In that case we can either add a new DROP_ON_COMPILE action or take the previous version of this diff.

Commit:
Update on "[JIT] preserve ignored function return value type"


Preserve the type of @ignore'd functions on serialization. Currently we first compile an @ignore'd function with it's annotated type when first compiling, but do not preserve it. This is important for being able to compile models with not-yet-supported features in JIT.

```
@torch.jit.ignore
def unsupported(x):
    return x

def foo():
   if not torch.jit._is_scripting():
      return torch.linear(...)
   else:
      return unsupported(...)
```

driazati commented on torch/csrc/jit/passes/python_print.cpp:
```diff
@@ -934,22 +934,17 @@ struct PythonPrintPass {
         }
-
-        if (value->ignore_on_export) {
-          stmt << "ops.prim.IgnoredPythonOp";
```
There's an implementation for this in `register_prim_ops` that should also be deleted

eellison replied:
Since serialized models might contain the op i dont think we shoudl remove it

driazati commented on torch/_jit_internal.py:
```diff
@@ -219,23 +220,22 @@ def unused_method(self, x):
 
 
-def ignore(drop_on_export=False):
+def ignore(drop=False, **kwargs):
```
The names could still use some work I think, `ignore` + `drop` isn't really clear what's going on. Maybe there should be two separate decorators for `@ignore` and `@ignore(True)` since their behavior is pretty different

eellison replied:
Sure, i don't love the names either. But it's the name that we've previously published; I think having too much churn probably isn't worth slightly better names.

suo replied:
Since we are changing the behavior, there will be churn either way. At least adding better names will make it nicer in the long run. I would prefer to make the names as clear as possible this time so we don't have to deal with confusion forever

eellison replied:
Okay - what are the better names ?

eellison replied:
i'm going to land soon if no one has a concrete suggestion 😛

suo replied:
@ignore should stay as is, and the new one should be like @unused (following C++, which…I guess)

driazati commented on test/test_jit.py:
```diff
@@ -13804,6 +13804,7 @@ def forward(self, x):
         self.assertTrue(out[0].is_cuda)
 
 
+    @suppress_warnings
```
Catch the warning and check its message

driazati commented on torch/_jit_internal.py:
```diff
@@ -290,24 +290,35 @@ def forward(self, x):
+    # for backwards compat
+    drop_on_export = kwargs.pop("drop_on_export", None)
+    if drop_on_export:
+        warnings.warn("ignore(drop_on_export=True) has been deprecated. TorchScript will now drop the drop "
```
Can you make the category `DeprecationWarning`?

driazati commented on torch/_jit_internal.py:
```diff
@@ -290,24 +290,35 @@ def forward(self, x):
+        warnings.warn("ignore(drop_on_export=True) has been deprecated. TorchScript will now drop the drop "
+                      "call on compilation. Use drop=True now. {}")
+
+        warnings.warn("")
```
Empty warning?

Commit:
Update on "[JIT] preserve ignored function return value type"


Preserve the type of @ignore'd functions on serialization. Currently we first compile an @ignore'd function with it's annotated type when first compiling, but do not preserve it. This is important for being able to compile models with not-yet-supported features in JIT.

```
@torch.jit.ignore
def unsupported(x):
    return x

def foo():
   if not torch.jit._is_scripting():
      return torch.linear(...)
   else:
      return unsupported(...)
```

Commit:
Update on "[JIT] preserve ignored function return value type"


Preserve the type of @ignore'd functions on serialization. Currently we first compile an @ignore'd function with it's annotated type when first compiling, but do not preserve it. This is important for being able to compile models with not-yet-supported features in JIT.

```
@torch.jit.ignore
def unsupported(x):
    return x

def foo():
   if not torch.jit._is_scripting():
      return torch.linear(...)
   else:
      return unsupported(...)
```

jamesr66a commented:
It looks like this broke `pytorch_linux_xenial_py2_7_9_test`

suo commented:
reverting

Commit:
Update on "[JIT] preserve ignored function return value type"


Preserve the type of @ignore'd functions on serialization. Currently we first compile an @ignore'd function with it's annotated type when first compiling, but do not preserve it. This is important for being able to compile models with not-yet-supported features in JIT.

```
@torch.jit.ignore
def unsupported(x):
    return x

def foo():
   if not torch.jit._is_scripting():
      return torch.linear(...)
   else:
      return unsupported(...)
```

Commit:
Update on "[JIT] preserve ignored function return value type"


Preserve the type of @ignore'd functions on serialization. Currently we first compile an @ignore'd function with it's annotated type when first compiling, but do not preserve it. This is important for being able to compile models with not-yet-supported features in JIT.

```
@torch.jit.ignore
def unsupported(x):
    return x

def foo():
   if not torch.jit._is_scripting():
      return torch.linear(...)
   else:
      return unsupported(...)
```

Commit:
Update on "[JIT] preserve ignored function return value type"


Preserve the type of @ignore'd functions on serialization. Currently we first compile an @ignore'd function with it's annotated type when first compiling, but do not preserve it. This is important for being able to compile models with not-yet-supported features in JIT.

```
@torch.jit.ignore
def unsupported(x):
    return x

def foo():
   if not torch.jit._is_scripting():
      return torch.linear(...)
   else:
      return unsupported(...)
```

Commit:
Update on "[JIT] preserve ignored function return value type"


Preserve the type of @ignore'd functions on serialization. Currently we first compile an @ignore'd function with it's annotated type when first compiling, but do not preserve it. This is important for being able to compile models with not-yet-supported features in JIT.

```
@torch.jit.ignore
def unsupported(x):
    return x

def foo():
   if not torch.jit._is_scripting():
      return torch.linear(...)
   else:
      return unsupported(...)
```

Differential Revision: [D17272443](https://our.internmc.facebook.com/intern/diff/D17272443)

Commit:
Update on "[JIT] preserve ignored function return value type"


Preserve the type of @ignore'd functions on serialization. Currently we first compile an @ignore'd function with it's annotated type when first compiling, but do not preserve it. This is important for being able to compile models with not-yet-supported features in JIT.

```
@torch.jit.ignore
def unsupported(x):
    return x

def foo():
   if not torch.jit._is_scripting():
      return torch.linear(...)
   else:
      return unsupported(...)
```

Differential Revision: [D17272443](https://our.internmc.facebook.com/intern/diff/D17272443)

eellison commented:
https://github.com/pytorch/pytorch/pull/25263 landed this - i think ghimport got confused because this was landed and reverted.
